myset.h

```diff
@@ -0,0 +1,156 @@
+//
+// Created by Kinstintin on 28.04.2024.
+//
+
+#ifndef UNTITLED_MYSET_H
+#define UNTITLED_MYSET_H
+#include <random>
+#include <unordered_map>
+
+static std::uniform_int_distribution<int> uniform(1, INT32_MAX);
+std::random_device dev;
+static std::mt19937_64 rng(dev());
+
+
+template <typename T,
+        typename Compare = std::less<T>>
+class Myset{
+private:
+    struct Node{
+        Node* par;
+        Node* minim;
+        Node* left;
+        Node* right;
+        T key;
+        int prior;
+        int count;
+        explicit Node(const T &value){
+            key = value;
+            count = 1;
+            left = nullptr;
+            right = nullptr;
+            minim = this;
+            par = nullptr;
+            prior = uniform(rng);
+        }
+    };
+    void renew(Node* &n){
+        if (!n)
+            return;
+
+        n->minim = n->left != nullptr ? n->left->minim : n;
+        if (n -> left)
+            n -> left -> par = n;
+        if (n -> right)
+            n -> right -> par = n;
+    }
+    std::tuple <Node*, Node*> split(Node *tree,const T &key){
+        Node *left = nullptr, *right = nullptr;
+        if (!tree)
+            return std::make_tuple(left, right);
+        auto cmp = Compare();
+        if (cmp(tree -> key, key)) { // tree->key < key
+            left = tree;
+            std::tie(left -> right, right) = split(tree->right, key);
+        } else{ // tree->key > key
+            right = tree;
+            std::tie(left, right->left) = split(tree -> left, key);
+        }
+        renew(left);
+        renew(right);
+        if (left)
+            left -> par = nullptr;
+        if (right)
+            right -> par = nullptr;
+        return std::make_tuple(left, right);
+    }
+    Node* merge(Node *t1, Node *t2){
+        if (!t1)
+            return t2;
+        if (!t2)
+            return t1;
+
+        if (t1 -> prior > t2 -> prior){
+            t1 -> right = merge(t1 -> right, t2);
+            renew(t1);
+            return t1;
+        }
+        t2 -> left = merge(t1, t2 -> left);
+        renew(t2);
+        return t2;
+    }
+    void push_min(Node *n){
+        if  (!n)
+            return;
+        n-> minim = n -> left -> minim;
+        push_min(n -> par);
+    };
+    void replace_with_right(Node *t){ // consider right exists
+
+        if (t -> par){
+            t -> par -> left = t -> right;
+            t -> right -> par = t -> par;
+        } else{
+            root = t -> right;
+            root -> par = nullptr;
+        }
+    }
+    void dealloc_entity(Node *n){
```

Comment:
Maybe you should use destructor

Reply:
dealloc_entity removes value from hash table. I am not sure that this logic should be in Node destructor for incapsulation reasons

Reply:
Ok